File: Dynamic programming/leetcode_55.cpp

// Problem link ===>> https://leetcode.com/problems/jump-game/description/

#include <bits/stdc++.h>
#define ll long long
using namespace std;

class Solution
{
public:
    bool canJump(vector<int> &nums)
    {
        int gas = 0;
        for (int i = 0; i < nums.size(); i++)
        {
            if (gas < i)
                return false;
            gas = max(gas, i + nums[i]);
        }
        return true;
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cout.tie(nullptr);
    cin.tie(nullptr);

    Solution solution;

    vector<int> nums1 = {2, 3, 1, 1, 4};
    cout << solution.canJump(nums1) << endl;

    vector<int> nums2 = {3, 2, 1, 0, 4};
    cout << solution.canJump(nums2) << endl;

    return 0;
}

File: Dynamic programming/leetcode_55.py

# Problem link ===>>  https://leetcode.com/problems/jump-game/description/
# submission link == =>> https://leetcode.com/problems/jump-game/submissions/1557315580/


from typing import List

class Solution:
    def canJump(self, nums: List[int]) -> bool:
        gas = 0
        for i, n in enumerate(nums):
            if gas < i:
                return False
            gas = max(gas, i + n)
        return True


if __name__ == "__main__":
    solution = Solution()

    nums1 = [2, 3, 1, 1, 4]
    print(solution.canJump(nums1))

    nums2 = [3, 2, 1, 0, 4]
    print(solution.canJump(nums2))
